Components: Enhance Notice actions to allow more props like disabled and onClick with url

manzoorwanijk opened the pull request:
Related to #74090

## What?

Enhances the `Notice` component's action buttons to support additional `Button` component features.

## Why?

Notice action buttons were missing common Button features like `disabled`, etc. The `openInNewTab` prop existed in types but wasn't implemented, and `onClick` was ignored when `url` was provided.

## How?

- Updated `NoticeAction` types to allow `disabled` and `onClick` alongside `url`
- Remove `openInNewTab` from the types.

## Testing Instructions

1. Check out this branch
2. Run `npm run storybook:dev`
3. Navigate to Components > Feedback > Notice
4. Test the new stories:
   - ** With Disabled Action**: Verify the disabled button cannot be clicked
   - ** With Open In New Tab**: Verify link opens in new tab

## Testing Instructions for Keyboard

- Tab to action buttons and verify they receive focus
- Press Enter/Space to activate buttons
- Disabled buttons should still be focusable via keyboard

## Screenshots

<img width="296" height="125" alt="Screenshot 2025-12-18 at 2 59 46 PM" src="https://github.com/user-attachments/assets/0b8824da-ea93-4e34-a982-e3cb32d4aead" />

<img width="264" height="104" alt="Screenshot 2025-12-28 at 11 59 16 AM" src="https://github.com/user-attachments/assets/6f7213c4-983b-4c25-b884-450571b09498" />

manzoorwanijk committed:
Components: Enhance Notice action buttons with missing Button props

manzoorwanijk committed:
Components: Add stories for enhanced Notice action features

manzoorwanijk committed:
Add unit tests for actions

manzoorwanijk committed:
Update CHANGELOG.md

Mamaduka commented:
I wonder if some of this was intentional to keep UI more or less consistent.

gmjuhasz commented on packages/components/src/notice/stories/index.story.tsx:
```diff
@@ -141,3 +142,145 @@ export const NoticeListSubcomponent: StoryFn< typeof NoticeList > = () => {
+	const [ clickCount, setClickCount ] = useState( 0 );
+
+	return (
+		<>
```
Do we need this extra Wrapper?

manzoorwanijk replied:
Not anymore. I initially had the count text outside the notice and then I moved it inside but forgot to get rid of the fragment. Thanks

gmjuhasz commented on packages/components/src/notice/stories/index.story.tsx:
```diff
@@ -141,3 +142,145 @@ export const NoticeListSubcomponent: StoryFn< typeof NoticeList > = () => {
+					},
+				] }
+			>
+				onClick now works with url. Click count: { clickCount }
```
Nitpick: This is fixed now, but in the feature this sentence could make less sense in the story, should we just have "Click count"?

manzoorwanijk replied:
True. Let me update it.

gmjuhasz reviewed (commented):
Coming from Jetpack I also did find this Notice component lacking this functionality before, we even had our own Notice component to battle this. I think these changes still keep it simple, but a lot more versatile.

manzoorwanijk committed:
Clean up the story for click count

manzoorwanijk committed:
Merge branch 'trunk' into enhance/notice-component-actions

aduth commented on packages/components/src/notice/index.tsx:
```diff
@@ -123,6 +123,8 @@ function Notice( {
 									noDefaultClasses = false,
 									onClick,
 									url,
+									openInNewTab,
```
I think I'm generally on board with the idea of notice actions basically being a pass-through to button, but I'm less convinced about this unique property. Since along the same line of thinking, why would `Notice` actions support this and not `Button` , if they're basically the same concept. My sense is that the consumer could be assigning this via `target` as expecting to interface with Button.

manzoorwanijk replied:
I agree on that. It's there because the action accepts `url` instead of `href`. May be it's better to simply remove it regardless.

manzoorwanijk replied:
Removed that. I went ahead and also added support for `href` prop along with URL to make it more consistent. The consumer can pass one of the two but not both.

manzoorwanijk committed:
Merge branch 'trunk' into enhance/notice-component-actions

manzoorwanijk committed:
Remove openInNewTab prop in favour of the native "target" prop

manzoorwanijk committed:
Add support for 'href' prop

Notice actions now accept an 'href' property in addition to 'url', allowing more flexible anchor rendering.

manzoorwanijk committed:
Fix the type for openInNewTab used by snackbar component

aduth commented on packages/components/src/notice/types.ts:
```diff
@@ -3,27 +3,53 @@
-// `onClick` will be ignored.
-type NoticeActionWithURL = CommonNoticeActionProps & {
-	url: string;
-	openInNewTab?: boolean;
```
Hm, maybe I hadn't realized that this was a prop we already supported. I thought were adding something new. We probably shouldn't break support for that. Although I think it could be deprecated / discouraged in favor of `target`.

On a related note, we'll want to update the documented props for actions, and ideally explain how additional props for Button are supported:

https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/notice/README.md#actions-arraynoticeaction

manzoorwanijk replied:
As mentioned in the PR description, `openInNewTab` was only defined as a type but not implemented in the component, so it was unused anyway.

aduth replied:
Interesting! Sorry I missed that in your prior comment.

aduth commented on packages/components/src/notice/types.ts:
```diff
@@ -3,27 +3,53 @@
+	AnchorProps & {
+		href: string;
+		url?: never;
+	};
```
Is all this extra stuff so that we enforce that someone has to pass one or the other of `url` and `href` but not both? I'm not sure it's worth the extra complexity if so. I'm also curious if we could use `ButtonProps` rather than both `ButtonAsButtonProps` and `ButtonAsAnchorProps`

manzoorwanijk replied:
Yes, it’s to ensure only one is passed when the consumer uses TypeScript. I think it’s fine this way. The current behaviour is retained and it won’t break anything.

aduth commented on packages/components/src/notice/test/index.tsx:
```diff
@@ -120,4 +122,164 @@ describe( 'Notice', () => {
 	} );
+
+	describe( 'actions', () => {
+		it( 'should render a disabled action button', () => {
```
These tests might be a bit excessive in testing for all the different supported properties. Really all we care about is that props are passed through, right? So that's 1 test? At least for the ones where we're just looking at what happens with the button. It makes sense for the `should call onClick` regression test to be its own thing.

Same could said for the stories.

manzoorwanijk replied:
Those tests and stories can act as a documentation to showcase how it works?

aduth replied:
I guess my concern is that this creates additional maintenance burden and potential for drift†, increases the runtime of tests, and doesn't really create much value over communicating that the actions can use any of the props from button and setting the expectation that they're passed through.

† On the topic of drift, there's already drift here because we only cover 5-6 of the available 15+ Button props.

manzoorwanijk replied:
Thanks. I have cleaned up the tests and stories to consolidate them.

manzoorwanijk committed:
Consolidate tests and stories

manzoorwanijk committed:
Merge branch 'trunk' into enhance/notice-component-actions

manzoorwanijk committed:
Merge branch 'trunk' into enhance/notice-component-actions

manzoorwanijk committed:
Make it non-controversial for now

manzoorwanijk commented:
As mentioned in https://github.com/WordPress/gutenberg/issues/74090#issuecomment-3690067578, we want to go with a non-controversial approach, that is what I have done for now.

manzoorwanijk committed:
Fix unit tests

mirka commented on packages/components/src/notice/index.tsx:
```diff
@@ -151,8 +154,12 @@ function Notice( {
 										variant={ computedVariant }
-										onClick={ url ? undefined : onClick }
+										onClick={ onClick }
+										disabled={ disabled }
+										accessibleWhenDisabled
```
Nice 🧑‍🍳👌

mirka commented on packages/components/src/notice/stories/index.story.tsx:
```diff
@@ -141,3 +141,65 @@ export const NoticeListSubcomponent: StoryFn< typeof NoticeList > = () => {
+ * When both `url` and `onClick` are provided, the action renders as a link
+ * and `onClick` is still called when clicked.
+ */
+export const WithOnClickAndUrl: StoryFn< typeof Notice > = ( props ) => {
```
This story is probably fine to remove since we have a unit test, and it's something that should "just work" rather than a feature that needs to be highlighted.

mirka commented on packages/components/src/notice/types.ts:
```diff
@@ -4,23 +4,46 @@
 import type { MouseEventHandler, ReactNode } from 'react';
 
 type CommonNoticeActionProps = {
+	/**
+	 * The label for the action button.
+	 */
```
Thanks for adding prop descriptions!

mirka commented on packages/components/src/snackbar/types.ts:
```diff
@@ -56,10 +56,11 @@ export type SnackbarProps = Pick<
-			NoticeAction,
-			'label' | 'url' | 'onClick' | 'openInNewTab'
-		>[];
+		actions?: Array<
+			Pick< NoticeAction, 'label' | 'url' | 'onClick' > & {
+				openInNewTab?: boolean;
+			}
```
Ugh I'm conflicted about this API divergence between Snackbar and Notice now. Snackbar added the feature through a `openInNewTab` prop, and swaps the `Button` for an `ExternalLink` in that case (#69905).

Actually I think that might be preferable to supporting `rel` and `target` directly, because one of the guidelines of the design system is to have consistent cues (visual and accessible) on links that open new tabs. `ExternalLink` has those cues, while `Button` does not.

The problem is that `Snackbar` only has to support `variant="link"` buttons, while `Notice` actions have a wider API surface that conflicts with `ExternalLink`, so the TypeScript/documentation complexity is probably not worth the trouble to make it work polymorphically.

This alone may be a good reason to [support components directly](https://github.com/WordPress/gutenberg/issues/74090#issuecomment-3681296174) in `actions`, for the `@wordpress/ui` version of `Notice`.

Sorry @manzoorwanijk, this aspect was a bit more complicated that expected. We can probably merge this PR faster if we punt this `rel`/`target` forwarding part out of scope 😅 (The type fix here for `openInNewTab` where you moved it out from Notice and into Snackbar is good though.)

manzoorwanijk replied:
The issue here was that `openInNewTab` was only defined in the type but not implemented. So, it was unused anyway.

So, I replaced it with target and rel to have better control.

As you rightly said, here we have to support different variants and thus it makes more sense to have it this way.

mirka replied:
The problem is that:

- We don't really want to allow tab-opening links without the visual/accessible cues like `ExternalLink` has.
- It's not good that `Snackbar` actions and `Notice` actions have different APIs to achieve the same thing.

We can solve these cleanly in a new API in `@wordpress/ui`, but it's hard to do this in `@wordpress/components`.

manzoorwanijk replied:
So what do we want to do here? Maybe we can close this PR, live with what we have for now, and hope to resolve the issues in `@wordpress/ui`?

SirLouen replied:
> It's not good that Snackbar actions and Notice actions have different APIs to achieve the same thing.

Why not replicate then? It shouldn't be too difficult. Maybe something like the Cover Block `cover/shared.js` could work here?

manzoorwanijk replied:
I have updated the implementation to use `ExternalLink` when `openInNewTab` is `true`.

<img width="264" height="104" alt="Screenshot 2025-12-28 at 11 59 16 AM" src="https://github.com/user-attachments/assets/6f7213c4-983b-4c25-b884-450571b09498" />

mirka replied:
Sorry if I was unclear. The main reason it's hard to do this in `@wordpress/components` is the type/documentation complexity it adds:

> The problem is that `Snackbar` only has to support `variant="link"` buttons, while `Notice` actions have a wider API surface that conflicts with `ExternalLink`, so the TypeScript/documentation complexity is probably not worth the trouble to make it work polymorphically.

For example we now have this discriminated union:

```ts
export type NoticeAction = NoticeActionWithURL | NoticeActionWithoutUrl;
```

The components squad was initially quite enthusiastic to do this kind of typing, but in practice, it gave us a lot more trouble than we imagined 😅 It can cause difficulty for consumers, especially those not advanced in TS, when they have to reuse these types for their own components. And no matter how feasible the TS representation is, it's always complicated to translate that into human-facing documentation.

And there are already some inaccuracies in the types, even with the attempted discriminated union, because certain props are not valid for the `ExternalLink` case (e.g. `disabled`, `noDefaultClasses`, `variant`). Just a lot of opportunities for human error and confusion to creep in.

So although I wouldn't go as far as to say these are "blockers", it does make me weigh the cost–benefit critically. It's almost like adding tech debt to a component that's about to be sunsetted. @ciampo or @aduth may have a second opinion.

manzoorwanijk replied:
I agree that those complicated types can sometimes be difficult to deal with, but I am still not sure what we want to do here?

Should we instead allow `Notice` `actions` composition by allowing it to be a `React.ReactNode` along with the current array structure, and then render it directly?

```jsx
<div className="components-notice__actions">
	{ Array.isArray( actions ) ? actions.map(...) : actions }
</div>
```

mirka replied:
Like Marco said in https://github.com/WordPress/gutenberg/issues/74090#issuecomment-3681251227, "quick wins and fixes" are fine, but we don't want to introduce new features that add maintenance/migration cost to these soon-to-be-deprecated componentry.

The `onClick` and `disabled` are quick fixes, but `openInNewTab` turned out to be on the level of a "new feature" that does indeed add to the future cost. So I'd prefer we remove the `openInNewTab` feature from the scope of this PR, given that we can support it cleaner in the new `@wordpress/ui` package.

manzoorwanijk replied:
`openInNewTab` was already defined there (in types) before this PR but was not implemented.

mirka replied:
Yes, it's messed up but let's leave it as is, unimplemented. I haven't looked into whether the types can be corrected easily so a `Notice` action won't accept an `onInNewTab` property, but that doesn't have to be done in this PR if it isn't simple.

manzoorwanijk replied:
Done. Thanks.

manzoorwanijk replied:
The types for `Notice` are now clean and straightforward.

manzoorwanijk commented on packages/components/src/notice/stories/index.story.tsx:
```diff
@@ -141,3 +141,65 @@ export const NoticeListSubcomponent: StoryFn< typeof NoticeList > = () => {
+			Click count: { clickCount }
+		</Notice>
+	);
+};
```
```suggestion
```

manzoorwanijk committed:
Remove unnecessary story

SirLouen commented:
> So what do we want to do here? Maybe we can close this PR, live with what we have for now, and hope to resolve the issues in @wordpress/ui?

@manzoorwanijk my conclusion reading through the whole convo is the following: limit the scope of this PR, by removing the open in new tab elements (alas, we missed `rel` & `target`). You will be able to push at least something in this PR, but not entirely.

manzoorwanijk committed:
Force openInNewTab to render ExternalLink

Replaces the 'target' and 'rel' props with a boolean 'openInNewTab' prop for Notice action buttons, which, if true, forces to render the ExternalLink

manzoorwanijk committed:
Restore Snackbar types

manzoorwanijk commented:
I have updated the PR (along with the description) to reduce the scope to do only three things:
- Allow `disabled` prop for buttons.
- Allow `onClick` alongside `url`.
- Wire up `openInNewTab` to render `ExternalLink` when it's `true`.

manzoorwanijk committed:
Move onClick to common props type and improve type name

manzoorwanijk committed:
Update docs

mirka commented:
I'm afraid I can't enthusiastically greenlight the `openInNewTab` part, but everything else is good on my end.

manzoorwanijk committed:
Revert "Restore Snackbar types"

This reverts commit a96c41539b168f432120a9015259bdb3647df229.

manzoorwanijk committed:
Limit the scope to `disabled` and `onClick` prop

manzoorwanijk committed:
Simplify types

manzoorwanijk committed:
Remove unused types

manzoorwanijk committed:
Clean up snackbar types

mirka reviewed (approved):
Looks good, thanks for your patience here 🙏
